Libraries/LibCore/EventLoopImplementationWindows.cpp

```diff
@@ -176,47 +298,95 @@ static int notifier_type_to_network_event(NotificationType type)
+    NTSTATUS status = ntdll.NtCancelWaitCompletionPacket(notifier_data->wait_packet.handle, FALSE);
+    VERIFY(NT_SUCCESS(status));
+    notifiers.remove(&notifier);
+    // TODO: Reuse the data structure
 }
 
 intptr_t EventLoopManagerWindows::register_timer(EventReceiver& object, int milliseconds, bool should_reload)
 {
     VERIFY(milliseconds >= 0);
-    // FIXME: This is a temporary fix for issue #3641
-    bool manual_reset = static_cast<Timer&>(object).is_single_shot();
-    HANDLE timer = CreateWaitableTimer(NULL, manual_reset, NULL);
-    VERIFY(timer);
+    auto* thread_data = ThreadData::the();
+    VERIFY(thread_data);
+    auto& timers = thread_data->timers;
+
+    Threading::MutexLocker lock(thread_data->mutex);
+
+    auto timer_data = make<EventLoopTimer>();
+    timer_data->type = CompletionType::Timer;
+    timer_data->timer.handle = CreateWaitableTimer(NULL, FALSE, NULL);
+    timer_data->owner = object.make_weak_ptr();
+    timer_data->is_periodic = should_reload;
+    VERIFY(timer_data->timer.handle);
+
+    NTSTATUS status = ntdll.NtCreateWaitCompletionPacket(&timer_data->wait_packet.handle, GENERIC_READ | GENERIC_WRITE, NULL);
+    VERIFY(NT_SUCCESS(status));
 
     LARGE_INTEGER first_time = {};
     // Measured in 0.1μs intervals, negative means starting from now
-    first_time.QuadPart = -10'000 * milliseconds;
-    BOOL rc = SetWaitableTimer(timer, &first_time, should_reload ? milliseconds : 0, NULL, NULL, FALSE);
-    VERIFY(rc);
+    first_time.QuadPart = -10'000LL * milliseconds;
+    BOOL succeeded = SetWaitableTimer(timer_data->timer.handle, &first_time, should_reload ? milliseconds : 0, NULL, NULL, FALSE);
+    VERIFY(succeeded);
+
+    status = ntdll.NtAssociateWaitCompletionPacket(timer_data->wait_packet.handle, thread_data->iocp.handle, timer_data->timer.handle, timer_data.ptr(), NULL, 0, 0, NULL);
+    VERIFY(NT_SUCCESS(status));
 
-    auto& timers = ThreadData::the()->timers;
-    VERIFY(!timers.get(timer).has_value());
-    timers.set(Handle(timer), { object });
-    return reinterpret_cast<intptr_t>(timer);
+    auto timer_id = reinterpret_cast<intptr_t>(timer_data.ptr());
+    VERIFY(!timers.get(timer_id).has_value());
+    timers.set(timer_id, move(timer_data));
+    return timer_id;
 }
 
 void EventLoopManagerWindows::unregister_timer(intptr_t timer_id)
 {
-    if (ThreadData::the())
-        ThreadData::the()->timers.remove(reinterpret_cast<HANDLE>(timer_id));
+    if (auto* thread_data = ThreadData::the()) {
+        Threading::MutexLocker lock(thread_data->mutex);
+        auto maybe_timer = thread_data->timers.get(timer_id);
+        if (!maybe_timer.has_value())
+            return;
+        auto* timer = maybe_timer.value();
+        // NOTE: Should we be removing a signalled packet that has yet to be dequed or not? Currently we do, but it can be changed.
+        ntdll.NtCancelWaitCompletionPacket(timer->wait_packet.handle, TRUE);
+        thread_data->timers.remove(timer_id);
```

Comment:
Can we use `HashMap::find()` here to avoid a double lookup? (Actually, perhaps `HashMap::take()` is even more ideal.)

Also, is there a reason you suspect that removing a signaled timer would be a problem? I don't think it's reasonable for a timer to fire after it's been unregistered, I suspect the only way that may be _somewhat_ reasonable is if it's being unregistered from a different thread.